File: Csharp/Lab8/Lab8/Lab8/cng/Model/Collections/neww/queue/QueueImpl.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Text;

namespace Lab8.cng.Model.Collections.neww.queue
{
    public class QueueImpl<T> : AbstractCollection<T>, IQueueInterface<T>
    {
        private object[] _queue = new object[4];
        private int _sizeVar;
        private int _queueArraySize = 4;


        public QueueImpl()
        {
        }

        public QueueImpl(T[] ts)
        {
            foreach (T t in ts)
            {
                Enqueue(t);
            }
        }

        public void Enqueue(T t)
        {
            if (_sizeVar == _queueArraySize - 1)
            {
                Resize(_queueArraySize * 2);
            }

            _queue[_sizeVar++] = t;
        }


        public T Dequeue()
        {
            T t = (T) _queue[0];
            object[] objects = new object[_queue.Length - 1];
            for (int i = 0; i < _queue.Length - 1; i++)
            {
                objects[i] = _queue[i + 1];
            }
            _queue = objects;
            _sizeVar--;
            return t;
        }


        public T Peek()
        {
            return (T) _queue[0];
        }


        public override int Size()
        {
            return _sizeVar;
        }

        public override void Add(T t)
        {
            Enqueue(t);
        }


        public override int GetHashCode()
        {
            const int hashconst = 31;
            int result = 1;
            for (int i = 0; i < _sizeVar; i++)
            {
                result += result * hashconst + _queue[i].GetHashCode();
            }

            return result;
        }

        //iterator
        public override IEnumerator<T> GetEnumerator()
        {
            for (int i = 0; i < _sizeVar; i++)
            {
                yield return (T) _queue[i];
            }
        }
        IEnumerator IEnumerable.GetEnumerator()
        {
            return GetEnumerator();
        }
        //iterator

        public override bool Equals(Object obj)
        {
            return this.GetHashCode() == obj.GetHashCode() && obj.GetHashCode() == this.GetHashCode();
        }


        public override String ToString()
        {
            StringBuilder sb = new StringBuilder().Append("Content\n");
            for (int i = 0; i < _sizeVar; i++)
            {
                sb.Append(i).Append("\n");
            }

            return sb.ToString();
        }

        


        //PRIVATE METHODS

        private void Resize(int newLength)
        {
            _queueArraySize = newLength;
            object[] newArray = new object[newLength];
            for (int i = 0; i < _sizeVar; i++)
            {
                newArray[i] = _queue[i];
            }
            _queue = newArray;
        }
    }
}

File: Java/Desktop/Saod/Lab3/src/by/bntu/fitr/prostrmk/collections/QueueImpl.java
package by.bntu.fitr.prostrmk.collections;

import by.bntu.fitr.prostrmk.collections.queue.QueueInterface;

import java.util.Collection;
import java.util.Iterator;

public class QueueImpl<T> extends AbstractCollection<T> implements QueueInterface<T> {

    private Object []queue = new Object[4];
    private int sizeVar;
    private int queueArraySize = 4;


    public QueueImpl(){

    }

    public QueueImpl(T...ts){
        for (T t : ts) {
            enqueue(t);
        }
    }

    @Override
    public void enqueue(T t) {
        if (sizeVar == queueArraySize - 1){
            resize(queueArraySize * 2);
        }
        queue[sizeVar++] = t;
    }

    @Override
    public T dequeue() {
        T t = (T) queue[0];
        Object []objects = new Object[queue.length - 1];
        System.arraycopy(queue, 1, objects, 0, queue.length - 1);
        queue = objects;
        sizeVar--;
        return t;
    }

    @Override
    public T peek() {
        return (T) queue[0];
    }

    @Override
    public int size() {
        return sizeVar;
    }

    @Override
    protected void add(T t) {
        enqueue(t);
    }

    @Override
    public void addAll(Collection c) {
        for (Object o : c) {
            enqueue((T) o);
        }
    }

    @Override
    public Iterator<T> iterator() {
        return new Iterator<T>() {
            private int index = 0;
            @Override
            public boolean hasNext() {
                return index < sizeVar;
            }

            @Override
            public T next() {
                return (T) queue[index++];
            }
        };
    }

    @Override
    public int hashCode() {
        final int HASHCONST = 31;
        int result = 1;
        Iterator iterator = this.iterator();
        while(iterator.hasNext()){
            result +=  result * HASHCONST + iterator.next().hashCode();
        }
        return result;
    }

    @Override
    public boolean equals(Object obj) {
        return this.hashCode() == obj.hashCode() && obj.getClass() == this.getClass();
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder().append("Content\n");
        for (int i = 0; i < sizeVar; i++) {
            sb.append(i).append("\n");
        }
        return sb.toString();
    }

    //PRIVATE METHODS

    private void resize(int newLength) {
        queueArraySize = newLength;
        Object[] newArray = new Object[newLength];
        System.arraycopy(queue, 0, newArray, 0, sizeVar);
        queue = newArray;
    }
}
